include/stl2/view/generate_view.hpp

```diff
@@ -0,0 +1,130 @@
+
+#include <stl2/detail/fwd.hpp>
+#include <stl2/detail/concepts/function.hpp>
+#include <stl2/detail/concepts/object.hpp>
+#include <stl2/detail/iterator/unreachable.hpp>
+#include <stl2/detail/non_propagating_cache.hpp>
+#include <stl2/detail/semiregular_box.hpp>
+#include <stl2/detail/view/view_closure.hpp>
+#include <stl2/view/view_interface.hpp>
+#include <type_traits>
+#include <utility>
+
+STL2_OPEN_NAMESPACE {
+	namespace ext {
+		template<Invocable F>
+		requires CopyConstructibleObject<F>
+		struct generate_view : public view_interface<generate_view<F>>
+									, private detail::semiregular_box<F>
+									, private detail::non_propagating_cache<invoke_result_t<F&>> {
+		private:
+			using result_t = invoke_result_t<F&>;
+
+			template<bool IsConst>
+			struct __iterator;
+		public:
+			generate_view() = default;
+
+			constexpr generate_view(F f) noexcept(is_nothrow_move_constructible_v<F>)
+				: detail::semiregular_box<F>(std::move(f))
+			{}
+
+			constexpr auto begin() noexcept
+			{ return begin_impl(*this); }
+
+			constexpr unreachable end() const noexcept
+			{ return {}; }
+		private:
+			template<class Self>
+			static constexpr auto begin_impl(Self& self) noexcept
+			{ return __iterator<is_const_v<Self>>{self}; }
+
+			constexpr decltype(auto) as_box() noexcept
+			{ return static_cast<detail::semiregular_box<F>&>(*this); }
+
+			constexpr decltype(auto) as_cache() noexcept
+			{ return static_cast<detail::non_propagating_cache<result_t>&>(*this); }
+		};
+
+		template<class F>
+		template<bool IsConst>
+		struct generate_view<F>::__iterator {
+			using value_type = result_t;
+			using difference_type = ptrdiff_t; // I guess?
+			using iterator_category = input_iterator_tag;
+			using iterator_concept = input_iterator_tag;
+
+			__iterator() = default;
+
+			__iterator(generate_view& v)
+				: view_{&v}
+			{}
+
+			constexpr decltype(auto) operator*() noexcept
+			requires (!Invocable<F const>)
+			{ return deref_impl(*this); }
+
+			constexpr decltype(auto) operator*() const noexcept
+			requires Invocable<F const>
+			{ return deref_impl(*this); }
+
+			constexpr __iterator& operator++() noexcept
+			{
+				view_->as_cache().reset();
+				return *this;
+			}
+
+			constexpr __iterator operator++(int) noexcept
+			{
+				auto result = *this;
+				++(*this);
+				return result;
+			}
+
+			constexpr friend bool operator==(__iterator, unreachable) noexcept
+			{ return false; }
+
+			constexpr friend bool operator==(unreachable x, __iterator y) noexcept
+			{ return y == x; }
+
+			constexpr friend bool operator!=(__iterator x, unreachable y) noexcept
+			{ return !(x == y); }
+
+			constexpr friend bool operator!=(unreachable x, __iterator y) noexcept
+			{ return !(y == x); }
+		private:
+			generate_view* view_;
+
+			template<class Self>
+			static constexpr decltype(auto) deref_impl(Self&& self)
+			{
+				auto& view = *self.view_;
+				if (!view.as_cache()) {
+					view.as_cache() = view.as_box().get()();
```

Comment:
Since we know the cache is empty, we should use `emplace` here:
```suggestion
					view.as_cache().emplace(view.as_box().get()());
```